src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php

```diff
@@ -0,0 +1,119 @@
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * @author Kuba Werłos <werlos@gmail.com>
+ */
+final class PhpUnitDataProviderReturnTypeFixer extends AbstractPhpUnitFixer
+{
+    public function getDefinition(): FixerDefinitionInterface
+    {
+        return new FixerDefinition(
+            'The return type of PHPUnit data provider must be `iterable`.',
+            [
+                new CodeSample(
+                    '<?php
+class FooTest extends TestCase {
+    /**
+     * @dataProvider provideSomethingCases
+     */
+    public function testSomething($expected, $actual) {}
+    public function provideSomethingCases(): array {}
+}
+',
+                ),
+                new CodeSample(
+                    '<?php
+class FooTest extends TestCase {
+    /**
+     * @dataProvider provideSomethingCases
+     */
+    public function testSomething($expected, $actual) {}
+    public function provideSomethingCases() {}
+}
+',
+                ),
+            ],
+            'Data provider must return `iterable`, either an array of arrays or an object that implements the `Traversable` interface.',
+            'Risky when relying on signature of the data provider.',
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * Must run before ReturnTypeDeclarationFixer.
+     * Must run after CleanNamespaceFixer.
+     */
+    public function getPriority(): int
+    {
+        return 0;
+    }
+
+    public function isRisky(): bool
+    {
+        return true;
+    }
+
+    protected function applyPhpUnitClassFix(Tokens $tokens, int $startIndex, int $endIndex): void
+    {
+        $dataProviderAnalyzer = new DataProviderAnalyzer();
+        $functionsAnalyzer = new FunctionsAnalyzer();
+
+        foreach (array_reverse($dataProviderAnalyzer->getDataProviders($tokens, $startIndex, $endIndex)) as $dataProviderAnalysis) {
+            $typeAnalysis = $functionsAnalyzer->getFunctionReturnType($tokens, $dataProviderAnalysis->getNameIndex());
+
+            if (null === $typeAnalysis) {
+                $argumentsStart = $tokens->getNextTokenOfKind($dataProviderAnalysis->getNameIndex(), ['(']);
+                $argumentsEnd = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_PARENTHESIS_BRACE, $argumentsStart);
+
+                $tokens->insertAt(
+                    $argumentsEnd + 1,
+                    [
+                        new Token([CT::T_TYPE_COLON, ':']),
+                        new Token([T_WHITESPACE, ' ']),
+                        new Token([T_STRING, 'iterable']),
+                    ],
+                );
+
+                continue;
+            }
+
+            if ('iterable' === $typeAnalysis->getName()) {
+                continue;
+            }
+
+            $typeStartIndex = $tokens->getNextMeaningfulToken($typeAnalysis->getStartIndex() - 1);
+            $typeEndIndex = $typeAnalysis->getEndIndex();
+
+            if ($tokens->generatePartialCode($typeStartIndex, $typeEndIndex) !== $typeAnalysis->getName()) {
```

Comment:
```suggestion
            // @TODO: drop condition and it's body when PHP 8+ is required
            if ($tokens->generatePartialCode($typeStartIndex, $typeEndIndex) !== $typeAnalysis->getName()) {
```

Reply:
btw, @Wirone , this is example why you cannot downgrade tool to old PHP engine and claim it gonna work - even if it would execute, the tokenizer and syntax of file we want to fix is different under old PHP engine

Reply:
I don't get why - downgrading is only for replacing language features that are not supported on lower versions of the language. It does not change the logic or remove conditions. Both PHPstan and Rector are downgraded for release and works properly, even though some features are related to specific PHP versions that are higher than language version to which tool the is downgraded. Imagine the same codebase we have now, but using PHP8+ features, which after downgrade looks like... current code 🙂. Why wouldn't it work?

Reply:
you can write code in PHP7.2, but if logic of that code is not covering PHP 7.2 syntax, then the code will not run properly.

So, imagine we say that we support only 8.2, there are no more dedicated test cases for PHP 7.x and extra conditions for specific 7.x syntax (like the one here) - even if you downgrade our source code, it wouldn't automatically add such extra condition

Reply:
But this has nothing to do with downgrading core code, but just technical decision which version is supported. If you decide to support 7.4, then you keep conditions in the code and test cases for 7.4, and downgrade to this version too. If you drop support for 7.4, then you remove conditions, tests and you change downgrading rules to go down to 8.0 only. So these are separate (but loosely related) concepts.

It works like this: you code in 8.2 and have such requirement, but during release you downgrade the code, change Composer requirements (so it can be used on lower PHP versions), test again if everything works and tag new version. Downgrading process has nothing to do with how fixers are covering PHP versions.

Reply:
> But this has nothing to do with downgrading core code, but just technical decision which version is supported.

for this project it was always the same decision. As with every bump of "which version of PHP we use to write our code", we were always getting rid of lot of legacy staff we no longer needed (eg conditions for old PHP tokenized content).

dependency is also interesting part - take our current dependencies and try to install them on 7.2 or 5.x. and I don't want to have a mess to support wide matrix of dependency versions.

Reply:
overall, I'm bringing to your visibility the complexity of maintaining code for low PHP tokenizer - it's not only how we transform our (sugar) syntax to lower version, but also that we would need to keep logic for those old tokenizers and for old dependency version.
This project is already complex enough and I do not want to have that extra complexity on top.

If you do believe it's worth to put effort to make newest PHP CS Fixer compatible and executable with legacy PHP version, do as you like to ;)
but I would recommend them to upgrade to new PHP engine instead of putting effort to use cutting-edge Fixer version

Reply:
@keradus you're mixing things and I believe we don't talk about the same. Since it's theoretical discussion, I'll just point basic things:

- I never meant to support 7.2 or even PHP5. What I was talking about is supporting _exact same_ versions like currently, but doing it while writing code that uses features from highest available version (for example native union types instead of phpDoc). To put it explicitly: write code on 8.2, downgrade it to 8.0 and support 8.0-8.2 (and starting to support 8.3). I don't know why you're talking about legacy PHP versions 🤷‍♂️.
- I did not bring downgrading thing up in terms of Fixer. I said it's good to have 2-way transformations because then it's possible to _use Fixer as a downgrade tool_. It's a topic for separate discussion.